datafusion/physical-expr/src/aggregate/bit_and_or_xor.rs

```diff
@@ -0,0 +1,1103 @@
+                if remainder_bits & (1 << i) != 0 {
+                    bit_or = bit_or | *value;
+                }
+            });
+
+            Some(bit_or)
+        }
+    }
+}
+
+fn bit_xor<T>(array: &PrimitiveArray<T>) -> Option<T::Native>
+where
+    T: ArrowNumericType,
+    T::Native: BitXorImplementation<Output = T::Native> + ArrowNativeTypeOp,
+{
+    let null_count = array.null_count();
+
+    if null_count == array.len() {
+        return None;
+    }
+
+    let data: &[T::Native] = array.values();
+
+    match array.nulls() {
+        None => {
+            let bit_xor = data.iter().fold(T::default_value(), |accumulator, value| {
+                accumulator ^ *value
+            });
+
+            Some(bit_xor)
+        }
+        Some(nulls) => {
+            let mut bit_xor = T::default_value();
+            let data_chunks = data.chunks_exact(64);
+            let remainder = data_chunks.remainder();
+
+            let bit_chunks = nulls.inner().bit_chunks();
+            data_chunks
+                .zip(bit_chunks.iter())
+                .for_each(|(chunk, mask)| {
+                    // index_mask has value 1 << i in the loop
+                    let mut index_mask = 1;
+                    chunk.iter().for_each(|value| {
+                        if (mask & index_mask) != 0 {
+                            bit_xor = bit_xor ^ *value;
+                        }
+                        index_mask <<= 1;
+                    });
+                });
+
+            let remainder_bits = bit_chunks.remainder_bits();
+
+            remainder.iter().enumerate().for_each(|(i, value)| {
+                if remainder_bits & (1 << i) != 0 {
+                    bit_xor = bit_xor ^ *value;
+                }
+            });
+
+            Some(bit_xor)
+        }
+    }
+}
+
+// Bit and/Bit or/Bit xor aggregation can take Dictionary encode input but always produces unpacked
+// (aka non Dictionary) output. We need to adjust the output data type to reflect this.
+// The reason bit and/bit or/bit xor aggregate produces unpacked output because there is only one
+// bit and/bit or/bit xor value per group; there is no needs to keep them Dictionary encode
+fn bit_and_or_xor_aggregate_data_type(input_type: DataType) -> DataType {
+    if let DataType::Dictionary(_, value_type) = input_type {
+        *value_type
+    } else {
+        input_type
+    }
+}
+
+// returns the new value after bit_and/bit_or/bit_xor with the new values, taking nullability into account
+macro_rules! typed_bit_and_or_xor_batch {
+    ($VALUES:expr, $ARRAYTYPE:ident, $SCALAR:ident, $OP:ident) => {{
+        let array = downcast_value!($VALUES, $ARRAYTYPE);
+        let delta = $OP(array);
+        Ok(ScalarValue::$SCALAR(delta))
+    }};
+}
+
+// bit_and/bit_or/bit_xor the array and returns a ScalarValue of its corresponding type.
+macro_rules! bit_and_or_xor_batch {
+    ($VALUES:expr, $OP:ident) => {{
+        match $VALUES.data_type() {
```

Comment:
I thin you could probably use https://docs.rs/arrow/latest/arrow/macro.downcast_primitive_array.html to avoid some of this replication